File: src/lib/validation/templates.ts
import type { ProfileConfig, ValidationTemplate, EnabledChecks } from '@/lib/types/validation'
import type { ColumnMapping } from '@/lib/parsing/types'

/** All checks enabled -- used for templates and reset functionality */
export const DEFAULT_ENABLED_CHECKS: EnabledChecks = {
  range_check: true,
  missing_data: true,
  duplicate_rows: true,
  near_duplicate_kp: true,
  outliers_zscore: true,
  outliers_iqr: true,
  kp_gaps: true,
  monotonicity: true,
  cross_column: true,
  spike_detection: true,
  coordinate_sanity: true,
  event_listing: true,
  position_consistency: true,
  kp_drift: true,
  segment_continuity: true,
}

/** Common statistical thresholds shared by all default templates */
const COMMON_CONFIG = {
  zscore_threshold: 3.0,
  iqr_multiplier: 1.5,
  kp_gap_max: null,
  duplicate_kp_tolerance: 0.001,
  monotonicity_check: true,
  enabled_checks: { ...DEFAULT_ENABLED_CHECKS },
  kp_drift_tolerance: 0.01,
  max_segment_distance: 100,
} as const satisfies Omit<ProfileConfig, 'ranges' | 'kp_drift_severity' | 'segment_continuity_severity'>

/**
 * 4 domain QC pack definitions matching the Python backend exactly.
 * These are read-only system templates -- users cannot modify or delete them.
 */
export const DEFAULT_TEMPLATES: ValidationTemplate[] = [
  {
    id: 'pipeline-as-laid',
    name: 'Pipeline As-Laid QC',
    survey_type: 'as-laid',
    description: 'Tight KP and depth validation for as-laid pipeline surveys',
    expectedColumns: ['kp', 'dob', 'depth', 'easting', 'northing'],
    config: {
      ...COMMON_CONFIG,
      kp_gap_max: 0.01,
      duplicate_kp_tolerance: 0.0005,
      kp_drift_tolerance: 0.01,
      max_segment_distance: 50,
      kp_drift_severity: 'critical',
      segment_continuity_severity: 'warning',
      zscore_threshold: 2.5,
      ranges: {
        dob: { min: 0, max: 3, tolerance: 0.1 },
        depth: { min: 0, max: 300 },
        easting: { min: 100000, max: 900000 },
        northing: { min: 0, max: 10000000 },
      },
    },
    is_default: true,
  },
  {
    id: 'as-built-survey',
    name: 'As-Built Survey QC',
    survey_type: 'as-built',
    description: 'DOC and cross-column focused validation for construction surveys',
    expectedColumns: ['kp', 'doc', 'dob', 'depth', 'easting', 'northing'],
    config: {
      ...COMMON_CONFIG,
      kp_gap_max: 0.05,
      duplicate_kp_tolerance: 0.001,
      kp_drift_tolerance: 0.02,
      max_segment_distance: 100,
      kp_drift_severity: 'warning',
      segment_continuity_severity: 'critical',
      zscore_threshold: 2.0,
      ranges: {
        doc: { min: 0, max: 2, tolerance: 0.05 },
        dob: { min: 0, max: 5 },
        depth: { min: 0, max: 300 },
        easting: { min: 100000, max: 900000 },
        northing: { min: 0, max: 10000000 },
      },
    },
    is_default: true,
  },
  {
    id: 'pre-commissioning',
    name: 'Pre-Commissioning QC',
    survey_type: 'pre-comm',
    description: 'Event listing and position focused validation for pre-commissioning surveys',
    expectedColumns: ['kp', 'dob', 'depth', 'easting', 'northing', 'event_listing'],
    config: {
      ...COMMON_CONFIG,
      kp_gap_max: 0.1,
      duplicate_kp_tolerance: 0.002,
      kp_drift_tolerance: 0.05,
      max_segment_distance: 200,
      kp_drift_severity: 'warning',
      segment_continuity_severity: 'warning',
      zscore_threshold: 3.0,
      ranges: {
        dob: { min: 0, max: 10 },
        doc: { min: 0, max: 10 },
        depth: { min: 0, max: 500 },
        easting: { min: 100000, max: 900000 },
        northing: { min: 0, max: 10000000 },
      },
    },
    is_default: true,
  },
  {
    id: 'general-survey',
    name: 'General Survey QC',
    survey_type: 'general',
    description: 'General purpose survey with generous thresholds for all column types',
    expectedColumns: [],
    config: {
      ...COMMON_CONFIG,
      kp_drift_tolerance: 0.03,
      max_segment_distance: 150,
      ranges: {
        dob: { min: 0, max: 10 },
        doc: { min: 0, max: 10 },
        top: { min: -500, max: 500 },
        depth: { min: 0, max: 500 },
        elevation: { min: -500, max: 500 },
        easting: { min: 100000, max: 900000 },
        northing: { min: 0, max: 10000000 },
        latitude: { min: -90, max: 90 },
        longitude: { min: -180, max: 180 },
      },
    },
    is_default: true,
  },
]

/**
 * Look up a default template by its ID.
 */
export function getTemplateById(id: string): ValidationTemplate | undefined {
  return DEFAULT_TEMPLATES.find((t) => t.id === id)
}

/**
 * Suggest the best matching template based on column mappings.
 * Uses column overlap scoring: for each non-general pack, counts how many of its
 * expectedColumns are present in the mapped types. Returns the best match if
 * score >= 0.6, otherwise returns "general-survey".
 */
export function suggestProfile(mappings: ColumnMapping[]): string {
  const mappedTypes = new Set<string>(
    mappings
      .filter((m) => !m.ignored && m.mappedType !== null)
      .map((m) => m.mappedType as string)
  )

  let bestId = 'general-survey'
  let bestScore = 0

  for (const template of DEFAULT_TEMPLATES) {
    if (!template.expectedColumns || template.expectedColumns.length === 0) continue
    const matched = template.expectedColumns.filter((col) => mappedTypes.has(col)).length
    const score = matched / template.expectedColumns.length
    if (score > bestScore) {
      bestScore = score
      bestId = template.id
    }
  }

  return bestScore >= 0.6 ? bestId : 'general-survey'
}

/**
 * Suggest the best matching template based on raw header names.
 * Used in the pipeline workflow where ColumnMapping objects are not available.
 * Maps common header names to survey column types, then scores against packs.
 */
export function suggestProfileFromHeaders(headers: string[]): string {
  const HEADER_TYPE_MAP: Record<string, string> = {
    kp: 'kp', chainage: 'kp', 'km post': 'kp',
    dob: 'dob', 'depth of burial': 'dob',
    doc: 'doc', 'depth of cover': 'doc',
    depth: 'depth', 'water depth': 'depth',
    easting: 'easting', east: 'easting', e: 'easting',
    northing: 'northing', north: 'northing', n: 'northing',
    top: 'top', 'top of pipe': 'top',
    elevation: 'elevation', elev: 'elevation',
    latitude: 'latitude', lat: 'latitude',
    longitude: 'longitude', lon: 'longitude', lng: 'longitude',
    event: 'event_listing', 'event listing': 'event_listing', 'event_listing': 'event_listing',
  }

  const types = new Set<string>()
  for (const h of headers) {
    const normalized = h.toLowerCase().trim()
    const type = HEADER_TYPE_MAP[normalized]
    if (type) types.add(type)
  }

  let bestId = 'general-survey'
  let bestScore = 0

  for (const template of DEFAULT_TEMPLATES) {
    if (!template.expectedColumns || template.expectedColumns.length === 0) continue
    const matched = template.expectedColumns.filter((col) => types.has(col)).length
    const score = matched / template.expectedColumns.length
    if (score > bestScore) {
      bestScore = score
      bestId = template.id
    }
  }

  return bestScore >= 0.6 ? bestId : 'general-survey'
}

File: backend/app/services/templates.py
"""Default validation profile templates and config resolution helpers."""

from app.models.schemas import EnabledChecks, ProfileConfig, RangeThreshold


# ---------------------------------------------------------------------------
# Default templates keyed by slug
# ---------------------------------------------------------------------------
DEFAULT_TEMPLATES: dict[str, ProfileConfig] = {
    "pipeline-as-laid": ProfileConfig(
        ranges={
            "dob": RangeThreshold(min=0, max=3, tolerance=0.1),
            "depth": RangeThreshold(min=0, max=300),
            "easting": RangeThreshold(min=100000, max=900000),
            "northing": RangeThreshold(min=0, max=10000000),
        },
        kp_gap_max=0.01,
        duplicate_kp_tolerance=0.0005,
        monotonicity_check=True,
        kp_drift_tolerance=0.01,
        max_segment_distance=50.0,
        kp_drift_severity="critical",
        segment_continuity_severity="warning",
        zscore_threshold=2.5,
        iqr_multiplier=1.5,
        enabled_checks=EnabledChecks(),
    ),
    "as-built-survey": ProfileConfig(
        ranges={
            "doc": RangeThreshold(min=0, max=2, tolerance=0.05),
            "dob": RangeThreshold(min=0, max=5),
            "depth": RangeThreshold(min=0, max=300),
            "easting": RangeThreshold(min=100000, max=900000),
            "northing": RangeThreshold(min=0, max=10000000),
        },
        kp_gap_max=0.05,
        duplicate_kp_tolerance=0.001,
        monotonicity_check=True,
        kp_drift_tolerance=0.02,
        max_segment_distance=100.0,
        kp_drift_severity="warning",
        segment_continuity_severity="critical",
        zscore_threshold=2.0,
        iqr_multiplier=1.5,
        enabled_checks=EnabledChecks(),
    ),
    "pre-commissioning": ProfileConfig(
        ranges={
            "dob": RangeThreshold(min=0, max=10),
            "doc": RangeThreshold(min=0, max=10),
            "depth": RangeThreshold(min=0, max=500),
            "easting": RangeThreshold(min=100000, max=900000),
            "northing": RangeThreshold(min=0, max=10000000),
        },
        kp_gap_max=0.1,
        duplicate_kp_tolerance=0.002,
        monotonicity_check=True,
        kp_drift_tolerance=0.05,
        max_segment_distance=200.0,
        kp_drift_severity="warning",
        segment_continuity_severity="warning",
        zscore_threshold=3.0,
        iqr_multiplier=1.5,
        enabled_checks=EnabledChecks(),
    ),
    "general-survey": ProfileConfig(
        ranges={
            "dob": RangeThreshold(min=0, max=10),
            "doc": RangeThreshold(min=0, max=10),
            "top": RangeThreshold(min=-500, max=500),
            "depth": RangeThreshold(min=0, max=500),
            "elevation": RangeThreshold(min=-500, max=500),
            "easting": RangeThreshold(min=100000, max=900000),
            "northing": RangeThreshold(min=0, max=10000000),
            "latitude": RangeThreshold(min=-90, max=90),
            "longitude": RangeThreshold(min=-180, max=180),
        },
        kp_drift_tolerance=0.03,
        max_segment_distance=150.0,
        kp_drift_severity="warning",
        segment_continuity_severity="warning",
        enabled_checks=EnabledChecks(),
    ),
}


# ---------------------------------------------------------------------------
# Template metadata for frontend consumption
# ---------------------------------------------------------------------------
TEMPLATE_METADATA: list[dict] = [
    {
        "id": "pipeline-as-laid",
        "name": "Pipeline As-Laid QC",
        "survey_type": "pipeline",
        "description": "Tight KP and depth-of-burial checks for as-laid pipeline surveys. Critical KP drift detection.",
        "expectedColumns": ["kp", "dob", "depth", "easting", "northing"],
    },
    {
        "id": "as-built-survey",
        "name": "As-Built Survey QC",
        "survey_type": "as-built",
        "description": "DOC and cross-column focus for construction verification. Aggressive spike detection, critical segment continuity.",
        "expectedColumns": ["kp", "doc", "dob", "depth", "easting", "northing"],
    },
    {
        "id": "pre-commissioning",
        "name": "Pre-Commissioning QC",
        "survey_type": "pre-comm",
        "description": "Event listing and position consistency focus for pre-commissioning surveys. Looser depth tolerances.",
        "expectedColumns": ["kp", "dob", "depth", "easting", "northing", "event_listing"],
    },
    {
        "id": "general-survey",
        "name": "General Survey QC",
        "survey_type": "general",
        "description": "General-purpose pack with generous limits for all numeric column types. Suitable for any survey data.",
        "expectedColumns": [],
    },
]


# ---------------------------------------------------------------------------
# Config resolution: ProfileConfig -> (flat_config_dict, enabled_checks_dict)
# ---------------------------------------------------------------------------
def resolve_config(config: ProfileConfig) -> tuple[dict, dict]:
    """Convert a ProfileConfig to (flat_config_dict, enabled_checks_dict).

    flat_config_dict has keys like "dob_min", "dob_max", "zscore_threshold", etc.
    enabled_checks_dict has keys like "range_check", "missing_data", etc.
    """
    flat: dict = {}

    # Flatten ranges into {col_type}_min / {col_type}_max / {col_type}_tolerance keys
    for col_type, threshold in config.ranges.items():
        flat[f"{col_type}_min"] = threshold.min
        flat[f"{col_type}_max"] = threshold.max
        if threshold.tolerance > 0:
            flat[f"{col_type}_tolerance"] = threshold.tolerance

    # Scalar thresholds
    flat["zscore_threshold"] = config.zscore_threshold
    flat["iqr_multiplier"] = config.iqr_multiplier
    flat["kp_gap_max"] = config.kp_gap_max
    flat["duplicate_kp_tolerance"] = config.duplicate_kp_tolerance

    # Chain-aware check config
    flat["kp_drift_tolerance"] = config.kp_drift_tolerance
    flat["max_segment_distance"] = config.max_segment_distance
    flat["kp_drift_severity"] = config.kp_drift_severity
    flat["segment_continuity_severity"] = config.segment_continuity_severity

    # Enabled checks as plain dict
    enabled = config.enabled_checks.model_dump()

    return flat, enabled
